crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -537,11 +537,26 @@ where
         .chain_constants(best_block_hash)?
         .confirmation_depth_k();
 
-    let maybe_last_archived_block = find_last_archived_block(
-        client,
-        segment_headers_store,
-        best_block_number.saturating_sub(confirmation_depth_k.into()),
-    )?;
+    // Trying to get the "best block to archive" in both cases: regular and fast sync.
+    let mut best_block_to_archive = best_block_number + 1u32.into();
+    for distance in 1..=(confirmation_depth_k + 1) {
+        let block_number_candidate = best_block_number.saturating_sub(distance.into());
+        if client.hash(block_number_candidate).ok().flatten().is_some()
+            && client
+                .hash(block_number_candidate.saturating_sub(1u32.into()))
+                .ok()
+                .flatten()
+                .is_some()
```

Comment:
This kind of tricky code should have comments. Even with deep knowledge of the code I do not understand why you need to also check the parent just from reading the code.

Reply:
> // We might add the block on the fast sync and we need to check for the parent block
> // availability as well to continue iteration.

I am still missing a link between fast sync and parent block. I do not understand why fast sync automatically implies the need to check for parent block. In fact I expect we may not have it once we restart node after single block insertion at the beginning of the segment. There will not be a parent block in that case, but it is fine.

Reply:
Why? This is exactly the kind of technical debt that will live in the code for years because no one will understand why it was done this way and not in more straightforward way. I'd reather understand why something that should have worked didn't.